src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs

```diff
@@ -54,366 +57,1142 @@ internal static class ThreadPoolGlobals
+
+                                            // unlock other prev slot
+                                            // must be after we moved other enq to the next slot, or someone may pop prev and break continuity of full slots.
+                                            enqPrevSlot.SequenceNumber = prevSequenceNumber;
+                                            return result;
+                                        }
+
+                                        // failed to lock desired half-way slot.
+                                        // restore via CAS, in case target slot has been robbed to
+                                        Interlocked.CompareExchange(ref halfSlot.SequenceNumber, halfPosition + Full, halfPosition + Change);
+                                    }
+                                }
+
+                                // failed to lock actual enqueue end, restore with CAS, in case target slot has been robbed to/from
+                                Interlocked.CompareExchange(ref enqPrevSlot.SequenceNumber, prevSequenceNumber, prevSequenceNumber + Change);
+                            }
+                        }
+                    }
 
-                                    // Check for nulls in the array.
-                                    if (obj == null) continue;
+                    // "this" is a sentinel for a failed robbing attempt
+                    return this;
+                }
 
-                                    m_array[idx] = null;
-                                    return obj;
-                                }
-                                else
+                /// <summary>
+                /// Searches for the given callback and removes it.
+                /// Returns "true" if actually removed the item.
+                /// </summary>
+                internal bool TryRemove(object callback)
+                {
+                    for (int position = _queueEnds.Enqueue - 1, l = position - RemoveRange; position != l; position--)
+                    {
+                        ref Slot slot = ref this[position];
+                        if (slot.Item == callback)
+                        {
+                            // lock Dequeue (so that the slot would not be robbed while we are taking it)
+                            var deqPosition = _queueEnds.Dequeue;
+                            ref var deqSlot = ref this[deqPosition];
+                            if (Interlocked.CompareExchange(ref deqSlot.SequenceNumber, deqPosition + Change, deqPosition + Full) == deqPosition + Full)
+                            {
+                                // lock the slot.
+                                if (Interlocked.CompareExchange(ref slot.SequenceNumber, position + Change, position + Full) == position + Full)
                                 {
-                                    // Failed, restore head.
-             
```

Comment:
Not sure why it's showing the lines above for the comment, but I intended for this comment to be for [this line](https://github.com/dotnet/coreclr/pull/18403/files/61234fa4560a2319e36445556532727f856b9289#diff-b9b22fdb097c803c0ca6fa45655b24f7R1187) and surrounding code

Reply:
> numOutstandingThreadRequests that is read immediately following a successful Enqueue must be at least as recent as the writes that complete the Enqueue operation are made visible to other threads

That is exactly a concern. The `numOutstandingThreadRequests` being read after Enqueue would be in a worst case as old as the last CAS (since reads can "move" before writes, or because writes after CAS are still in the writebuffer, whichever way to look at it). The last CAS would be locking of the slot. 

That is countered by detecting inconsistent states in Dequeue and just claiming `missedSteal` - I.E. we could be responsible for dequeuing this item, but we came too early so we cannot do our job (unless we spin until consistent state, like in global queue). Instead we request that someone looks at the queues again. And if that someone finds an item in another queue, he will ask that someone else looks at the queues again, and so on. Eventually someone looks in this queue.
 
The sequence that you describe is a very good scenario. It should get resolved (in theory!) at:

Try to steal from local queue, queue found empty without performing any interlocked operations, no missed steal because enqueue index == dequeue index and **sequence number change from T1 is not seen**

The sequence numbers for local queues are CAS-ed into locked state, so that should be globally visible before we read `numOutstandingThreadRequests` on Enqueue. 
The incoming worker decrements `numOutstandingThreadRequests` via CAS so it will see the at least the locked state even on systems with read buffers. In such case the worker claims `missedSteal`

I am saying "in theory" because indeed there seems to be an issue. To support `Pop`, we lock the slot at `Enqueue-1`, but Dequeue will check the `Enqueue`. I used to lock the same slot before `Pop` was supported, but that changed.
I think what saves us is that write-after-read issues are very rare and `Enqueue-1` is likely to be on same cache line as `Enqueue`.

It is still an issue. The same as the old pool had. 
(I assumed slot locking removed it, but after supporting `Pop` it is back, at least in theory).

If asking for more threads is a sufficient mitigation, let's continue doing that.

Reply:
> The sequence numbers for local queues are CAS-ed into locked state, so that should be globally visible before we read numOutstandingThreadRequests on Enqueue.

Based on this line ([link](https://github.com/dotnet/coreclr/pull/18403/files#diff-b9b22fdb097c803c0ca6fa45655b24f7R775)), the sequence number of the previous slot is modified with an interlocked operation, but the sequence number of the slot where the new item is stored (the one I believe that matters for my point above) is updated via `Volatile.Write` ([link](https://github.com/dotnet/coreclr/pull/18403/files#diff-b9b22fdb097c803c0ca6fa45655b24f7R796)), and may be invisible to other threads for a while, so I still don't see how `missedSteal == true` is guaranteed in such a case.

> I am saying "in theory" because indeed there seems to be an issue. To support Pop, we lock the slot at Enqueue-1, but Dequeue will check the Enqueue. I used to lock the same slot before Pop was supported, but that changed.
I think what saves us is that write-after-read issues are very rare and Enqueue-1 is likely to be on same cache line as Enqueue.

I didn't follow that entirely, but would it be right to understand that you were previously locking the same slot into which the new work item would be stored? If that's the case it would solve the scenario above, but I would be willing to bet that another breaking scenario would exist. At some point it's not worth breaking a solution anymore; when there are enough problems, just do the safe thing :). I'm not sure we have reached that point, but it certainly makes me uncomfortable, especially because it's also totally understandable why such issues would be more difficult to repro with the changes.

> If asking for more threads is a sufficient mitigation, let's continue doing that.

+1!

Reply:
> I didn't follow that entirely, but would it be right to understand that you were previously locking the same slot into which the new work item would be stored

That is the right assumption. 
And that, IMO, in combination with reporting `missedSteal` would be enough to guarantee that we are not losing threads. Since it would gurantee that the particular enqueue, even if incomplete, would be visible to any thread that arrives after we read `numOutstandingThreadRequests`, since that read cannot move before the CAS.

At the time I started locking `Enqueue-1` this assumption is no longer there. 
It is pretty cool that you noticed that!!!

I looked at the current implementation and, in theory, it should have the same issue. Enq is updated by ordinary (not CAS) write before checking `numOutstandingThreadRequests`, which can move before that write. 

So, if asking for more threads worked to mitigate this before, it should continue working. Let's do that.

Reply:
> Since it would gurantee that the particular enqueue, even if incomplete, would be visible to any thread that arrives after we read numOutstandingThreadRequests, since that read cannot move before the CAS.

Yea I agree that would have been a way of solving the issue. Although, it's not as preferable for a dequeuer to notice that something changed by an enqueuer enough to warrant the dequeuer to request a thread, rather it would be preferable for the enqueuer to guarantee that if a dequeuer would not notice an enqueue, that the enqueuer would request a thread, as that would allow minimizing thread requests from `Dispatch` without any loss. Either way would work though from a correctness perspective.